promql/prettier/prettier.go

```diff
@@ -0,0 +1,533 @@
+package prettier
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/rulefmt"
+	"github.com/prometheus/prometheus/promql/parser"
+	"io/ioutil"
+	"reflect"
+	"regexp"
+)
+
+const (
+	// PrettifyRules for prettifying rules files along with the
+	// expressions in them.
+	PrettifyRules = iota
+	// PrettifyExpression for prettifying instantaneous expressions.
+	PrettifyExpression
+	// itemNotFound is used to initialize the index of an item.
+	itemNotFound = -1
+)
+
+// Prettier handles the prettifying and formatting operation over a
+// list of rules files or a single expression.
+type Prettier struct {
+	files      []string
+	expression string
+	Type       uint
+	Node       parser.Expr
+	pd         *padder
+}
+
+type padder struct {
+	indent      string
+	columnLimit int
+	// isNewLineApplied represents a new line was added previously, after
+	// appending the item to the result. It should be set to default
+	// after applying baseIndent.
+	isNewLineApplied bool
+	// containsGrouping represents whether a binary expression contains
+	// a grouping modifier.
+	containsGrouping bool
+	// immediateScalar represents whether the immediate child of a binary
+	// expression is a scalar.
+	immediateScalar         bool
+	multiArgumentCall       bool
```

Comment:
Does this work with recursive calls?

Reply:
It should work. The basic purpose here is, to see if the current can has more than 1 argument or not.

Reply:
```
label_replace(
   label_replace(
      label_replace(
          ...
      ),
      ...
   ),
   ...
)
```

where `...` is replaced by some values of the right type.

Reply:
This should be handled. However, I am test out and paste the exact answer here.

Reply:
Generally it might be better to look those things up dynamically instead of keeping global state, since it's easy to forget always updating the state when necessary.

Reply:
(This refers to all of the booleans here.)

Reply:
Got the point.

Reply:
The output for 
```
label_join(label_join(label_join(up{job="api-server",src1="a",src2="b",src3="c"}, "foo", ",", "src1", "src2", "src3"), "foo", ",", "src1", "src2", "src3"), "foo", ",", "src1", "src2", "src3")
```

came out to be
```
  label_join(
    label_join(
      label_join(
        up{job="api-server", src1="a", src2="b", src3="c"},
        "foo",
        ",",
        "src1",
        "src2",
        "src3"
      )    ,
      "foo",
      ",",
      "src1",
      "src2",
      "src3"
    )  ,
    "foo",
    ",",
    "src1",
    "src2",
    "src3"
  )

```

Reply:
I think this can be resolved now. **@slrtbtfs**

Reply:
Resolving this.